File: js/gui/GUIBuilder.js
/* Copyright STIFTELSEN SINTEF 2012
 * 
 * This file is part of the Tinia Framework.
 * 
 * The Tinia Framework is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * The Tinia Framework is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with the Tinia Framework.  If not, see <http://www.gnu.org/licenses/>.
 */

dojo.provide("gui.GUIBuilder");


dojo.require("dijit.layout.BorderContainer");
dojo.require("dijit.layout.ContentPane");
dojo.require("dijit.layout.TabContainer");
dojo.require("dijit.form.TextBox");
dojo.require("dojo.fx.Toggler");
dojo.require("dijit.form.DropDownButton");
dojo.require("dijit.TooltipDialog");
dojo.require("dijit.form.CheckBox");
dojo.require("dijit.form.ToggleButton");
dojo.require("dijit.form.RadioButton");
dojo.require("dijit.form.Select");
dojo.require("dijit.form.NumberSpinner");
dojo.require("dijit.form.HorizontalSlider");
dojo.require("dijit.form.VerticalSlider");
dojo.require("dijit.layout.TabContainer");
dojo.require("dijit.layout.TabController");
dojo.require("dijit.form.Button");


dojo.require("gui.CheckBoxWithLabel");
dojo.require("gui.VerticalLayout");
dojo.require("gui.HorizontalLayout");
dojo.require("gui.Label");
dojo.require("gui.GridLayout");
dojo.require("gui.RadioButtonWithLabel");
dojo.require("gui.Canvas");


dojo.declare("gui.GUIBuilder", null, {
    constructor: function(modelLib, renderListURL, localMode, urlHandler, renderList) {
        this._modelLib = modelLib;
        this._renderListURL = renderListURL;
        this._localMode = !!localMode;
        this._urlHandler = urlHandler;
	this._showRenderList = renderList;
    },
    
    buildGUI: function( root) {
        var element = null;
        if(!root) {
            console.log("ERROR");
            return null;
        }
        switch(root.type()) {
            case 'ElementGroup':
                element = this._makeElementGroup(root);
                break;
            case 'HorizontalLayout':
                element = this._makeHorizontalLayout(root);
                break;
            case 'VerticalLayout':
                element = this._makeVerticalLayout(root);
                break;
            case 'TextInput':
                element = this._addKeyValue(new dijit.form.TextBox(), root);
                break;
            case 'Grid':
                element = this._makeGridLayout(root);
                break;
            case 'PopupButton':
                element = this._makePopupButton(root);
                break;
            case 'Checkbox':
                element = this._addLabel(this._addKeyValue(new gui.CheckBoxWithLabel(), root, "checked", ["clicked", "change"]), 
                    root, 'label');
                break;
            case 'Label':
                element = this._addKeyValue(new gui.Label(), root, "value", "change");
                break;
            case 'ComboBox':
                element = this._makeComboBox(root);
                break;
            case 'RadioButtons':
                element = this._makeRadioButtons(root);
                break;
            case 'DoubleSpinBox':
            case 'SpinBox':
                element = this._makeSpinBox(root);
                break;
            case 'HorizontalSlider':
                element = this._makeHorizontalSlider(root);
                break;
            case 'VerticalSlider':
                element = this._makeVerticalSlider(root);
                break;
            case 'TabLayout':
                element = this._makeTabLayout(root);
                break;
            case 'Button':
                element = this._makeButton(root);
                break;
            case 'Canvas':
                element = new gui.Canvas({"key" : root.key(), 
                                          "renderlistKey" : root.renderlistKey(), 
                                           "boundingboxKey": root.boundingBoxKey(),
                                           "modelLib" : this._modelLib,
                                            "localMode" : this._localMode,
                                            "renderListURL" : this._renderListURL,
                                            "resetViewKey"  : root.resetViewKey(),
                                          "urlHandler" : this._urlHandler,
					  "showRenderList" : this._showRenderList,
                                            "scripts" : root.scripts()});
                
                break;
                
        }
        if(element) {
            this._addVisibilityToggles(element, root);
        }
        else {
            console.log("ERROR, type=" + root.type());
        }
        return element;
    },
    
    
    _makeElementGroup: function(root) {
        var heading = dojo.create("b");
        if(root.showValue()) {
            heading.innerHTML = this._modelLib.getValue(root.key());
            this._modelLib.addLocalListener(root.key(), function(key, value){
                heading.innerHTML = value;
            });
        }
        else {
            heading.innerHTML = this._modelLib.getAnnotation(root.key());
        }
        var container = new gui.VerticalLayout();
        dojo.style(heading, "padding-top", "5px");
        dojo.style(heading, "padding-bottom", "2px");
        dojo.style(heading, "padding-left", "0px");
        dojo.style(heading, "display", "block");
        container.domNode.appendChild(heading);
        var child = this.buildGUI(root.child());
        if(child) {
            container.addChild(child);
        }
        return container;
    },
    
    
    _makeButton : function(root) {
        var button = new dijit.form.Button();
        var label = "";
        if(root.showValue()) {
            this._modelLib.addLocalListener(root.key(), function(key, value) {
                button.set("label", value);
            });
            label = this._modelLib.getValue(root.key());
        }
        else {
            label = this._modelLib.getAnnotation(root.key());
        }
        button.set("label", label);
        button.on("click", dojo.hitch(this, function(event)  {
            this._modelLib.updateElement(root.key(), false);
            this._modelLib.updateElement(root.key(), true);
        }));
        return button;
    },
    
    
    _makeTabLayout : function(root) {
        var tabLayout = new dijit.layout.TabContainer({doLayout: false,
                                                        controllerWidget: "dijit.layout.TabController"
                                                       });
        
        for(var i = 0; i < root.length(); i++) {
            if(root.child(i)) {
                var element = null;
                if(root.child(i).child()) {
                    element = this.buildGUI(root.child(i).child());
                }

                var options = {
                    "title"      : this._modelLib.getAnnotation(root.child(i).key()),
                    "closeable" : false,
                    "selected"  : i==0
                    };
                options.content = element ? element.domNode : "";
                var contentPane = new dijit.layout.ContentPane(options);

                tabLayout.addChild(contentPane);
            }
            
        }
        var content = new gui.VerticalLayout();
        content.addChild(tabLayout);
        return content;
    },
    
    
    _makeSpinBox: function(root) {
        var spinBox = null;
        var style = "width: 7em;";
        if(this._modelLib.isConstrained(root.key())) {
            var max = this._modelLib.getMax(root.key());
            var min = this._modelLib.getMin(root.key());
            spinBox = new dijit.form.NumberSpinner({
                constraints : {
                    "max" : max,
                    "min" : min
                },
                "style" : style 
                
            });
        } else {
            spinBox = new dijit.form.NumberSpinner({"style" : style});
        }
        
        

        return this._addMaxMinListeners(this._addKeyValue(spinBox, root, "value", ["change", "click"]), root);
    },
    _makeHorizontalSlider: function(root) {

        var max = this._modelLib.getMax(root.key());
        var min = this._modelLib.getMin(root.key());
        var slider = new dijit.form.HorizontalSlider({
            "maximum" : max,
            "minimum" : min,
            "style"   : "width: 150px;"
        });
        return this._addMaxMinListeners(this._addKeyValue(slider, root, "value", "change" ), root);
    },
    
    _makeVerticalSlider: function(root) {

        var max = this._modelLib.getMax(root.key());
        var min = this._modelLib.getMin(root.key());
        var slider = new dijit.form.VerticalSlider({
            "maximum" : max,
            "minimum" : min,
            "style"   : "height: 150px;"
        });
        return this._addMaxMinListeners(this._addKeyValue(slider, root, "value", "change"), root);
    },
   
    _makeVerticalLayout: function(root) {
        var layout = new gui.VerticalLayout();

 
        for(var i = 0; i < root.length(); i++) {
            var widget = this.buildGUI(root.child(i));
            if(widget) {
                layout.addChild(widget);
            }
        }
        
        return layout;
    },
    
    _makeGridLayout : function(root) {
        var layout = new gui.GridLayout(root.height(), root.width());
        
        for(var i = 0; i < root.height(); i++) {
            for(var j = 0; j < root.width(); j++) {
                if(root.child(i,j)) {
                    var widget = this.buildGUI(root.child(i,j));
                    if(widget) {
                        layout.addChild(widget, i, j);
                    } else {
                        layout.addChild(new gui.EmptyWidget(), i, j);
                    }
                }
                else {
                    layout.addChild(new gui.EmptyWidget(), i, j);
                }
            }
        }
        return layout;
    },
    
    _makeHorizontalLayout : function(root) {
        var layout = new gui.HorizontalLayout();
        for(var i = 0; i < root.length(); i++) {
            var widget = this.buildGUI(root.child(i));
            if(widget) {
                layout.addChild(widget);
            }
        }
        return layout;
        
    },
    
    _makePopupButton: function(root) {
        var button = dijit.form.DropDownButton();
        button.set('label', root.showValue() ? this._modelLib.getValue(root.key()) :
            this._modelLib.getAnnotation(root.key()));
        
        if(root.showValue()) {
            this._modelLib.addLocalListener(root.key(), function(key, value) {
                button.set("label", value);
            });
        }
        
        var dialog = new dijit.TooltipDialog();
        button.set("dropDown", dialog);
        var child = this.buildGUI(root.child());
        if(child) {
            dialog.set('content', child);
        }
        return button;
    },
    
    _makeComboBox : function(root) {
        var restriction = this._modelLib.getRestriction(root.key());
        var options = [];
        for(var i = 0; i < restriction.length; i++) {
            options[i]=  {
                label: restriction[i], 
                value: restriction[i], 
                selected: false
            };
            if(this._modelLib.getValue(root.key())== options[i].value) {
                options.selected = true;
            }
        }
        var combobox = new dijit.form.Select({
            "name": root.key(), 
            "options" : options, 
            maxHeight:100, 
            selected: this._modelLib.getValue(root.key()), 
            value: this._modelLib.getValue(root.key())
        });
        return this._addKeyValue(combobox, root, "value", ["change", "blur"]);
    },
    
    _makeRadioButtons : function(root) {
        var layout = new gui.VerticalLayout();
        var restriction = this._modelLib.getRestriction(root.key());
        for(var i = 0; i < restriction.length; i++) {
            (dojo.hitch(this, function() { // For closure (need options and radiobutton)
                var options = {
                    name: "model."+root.key(), 
                    value: restriction[i],
                    label: restriction[i], 
                    checked : false
                };
                if(options.value == this._modelLib.getValue(root.key())) {
                    options.checked = true;
                }
                var radioButton = new gui.RadioButtonWithLabel(options);
                radioButton.on("change", dojo.hitch(this, function(value) {
                    if(value) {
                        this._modelLib.updateElement(root.key(), options.value);
                    }
                }));
                this._modelLib.addLocalListener(root.key(), function(key, value) {
                    if(value == options.value) {
                       
                        radioButton.set("checked", value);
                    }
                }, this);
                layout.addChild(radioButton);
            }))();
        }
        return layout;
    },
    
    
    
    
    _addKeyValue: function(widget, root, slot, event) {
        if(!slot) {
            slot = "value";
        }
        if(!event) {
            event = "change";
        }
        if(typeof event == "string") {
            event = [event];
        }
        if(root.showValue()) {
            
            for(var i = 0; i < event.length; i++) {
                var currentEvent = event[i];
                (dojo.hitch(this, function() {
                    widget.on(currentEvent, dojo.hitch(this, function(value) {
                        this._modelLib.updateElement(root.key(), widget.get(slot));
                    }));
                }))();
            }
            
            (dojo.hitch(this, function() {
                widget.on("keypress", dojo.hitch(this, function(event) {
                    if(event.keyCode == 13) {
                        this._modelLib.updateElement(root.key(), widget.get(slot));
                    }
                }));
            }))();

            
            
            this._modelLib.addLocalListener(root.key(), function(key, value) {
                widget.set(slot, value);
            }, this);
            
            widget.set(slot, this._modelLib.getValue(root.key()));
        } 
        else {
            widget.set(slot, this._modelLib.getAnnotation(root.key(), "en"));
        }
        return widget;
    },
    
    _addMaxMinListeners : function(widget, root) {
        this._modelLib.addLocalListener(root.key(), function(key, value) {
            var max = this._modelLib.getMax(key);
            var min = this._modelLib.getMin(key);
            widget.set("minimum", this._modelLib.getMin(key));
            widget.set("maximum", this._modelLib.getMax(key));
            widget.set("constraints", {"max" : max, "min": min} );
        }, this);
        return widget;
    },
    
    
    
    _addVisibilityToggles : function(widget, root) {
        if(root.visibilityKey() && root.visibilityKey().length > 0) {
            this._modelLib.addLocalListener(root.visibilityKey(), function(key, value) {
                if(root.visibilityKeyInverted() == value) {
                    dojo.style(widget.domNode, "display", "none");
                }
                else {
                    dojo.style(widget.domNode, "display", "block");
                }
            });
            if(this._modelLib.getValue(root.visibilityKey()) == root.visibilityKeyInverted()) {
                dojo.style(widget.domNode, "display", "none");
            }
        }
        if(root.enabledKey() && root.enabledKey().length > 0) {
            this._modelLib.addLocalListener(root.enabledKey(), function(key, value) {
                if(root.enabledKeyInverted() == value) {
                    widget.set("disabled", true);
                }
                else {
                    widget.set("disabled", false);
                }
            });
            if(this._modelLib.getValue(root.enabledKey()) == root.enabledKeyInverted()) {
                widget.set("disabled", true);
            }
            
        }
    },
    
    _addLabel: function(widget, root,  slot) {
        if(!slot) {
            slot = "label";
        }
        widget.set(slot, this._modelLib.getAnnotation(root.key()));
        return widget;
    }
});



dojo.declare("gui.EmptyWidget", [dijit._Widget], {});

File: src/qtobserver/GUIBuilder.cpp
/* Copyright STIFTELSEN SINTEF 2012
 * 
 * This file is part of the Tinia Framework.
 * 
 * The Tinia Framework is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * The Tinia Framework is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with the Tinia Framework.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "tinia/qtobserver/GUIBuilder.hpp"
#include "tinia/qtobserver/moc/TabWidgetChildren.hpp"
#include "tinia/qtobserver/moc/ComboBox.hpp"
#include "tinia/qtobserver/moc/RadioButtonGroup.hpp"
#include "tinia/qtobserver/moc/SpinBox.hpp"
#include "tinia/qtobserver/moc/CheckBox.hpp"
#include "tinia/qtobserver/moc/Button.hpp"
#include "tinia/qtobserver/moc/HorizontalSlider.hpp"
#include "tinia/qtobserver/moc/ElementGroup.hpp"
#include "tinia/qtobserver/moc/Canvas.hpp"
#include "tinia/qtobserver/moc/DoubleSpinBox.hpp"
#include "tinia/qtobserver/moc/VisibilityController.hpp"
#include "tinia/qtobserver/moc/EnabledController.hpp"
#include "tinia/qtobserver/moc/StringController.hpp"
#include "tinia/qtobserver/moc/PopupEventFilter.hpp"
#include "tinia/qtobserver/moc/FileDialogButton.hpp"
#include "tinia/qtobserver/impl/utils.hpp"
#include <QLayout>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QGridLayout>
#include <QSpacerItem>
#include <QSizePolicy>
#include <QLabel>
#include <QObject>
#include <QWidget>
#include <QPushButton>
using namespace tinia::qtobserver::impl;
namespace tinia {
namespace qtobserver {

GUIBuilder::GUIBuilder( std::shared_ptr<model::ExposedModel>   model,
                        jobobserver::Job*                       job,
                        QTObserver*                             observer,
                        const bool                              perf_mode,
                        QGLWidget*                              root_context )
   : m_model(model),
     m_job(job),
     m_observer( observer ),
     m_perf_mode( perf_mode ),
     m_root_context( root_context )
{
}


QWidget* GUIBuilder::buildGUI(model::gui::Element* root, QWidget *parent )
{
    QWidget* widget = NULL;

    if( widget == NULL ) {
        model::gui::TabLayout* tab_layout = dynamic_cast<model::gui::TabLayout*>( root );
        if( tab_layout != NULL ) {
            widget = addTabLayout( tab_layout, parent );
        }
    }
    if( widget == NULL ) {
        model::gui::Grid* grid = dynamic_cast<model::gui::Grid*>( root );
        if( grid != NULL ) {
            widget = addGrid( grid, parent );
        }
    }
    if( widget == NULL ) {
        model::gui::TextInput* text_input = dynamic_cast<model::gui::TextInput*>( root );
        if( text_input != NULL ) {
            widget = new QLineEdit( parent );
            new StringController( widget,
                                  m_model,
                                  text_input->key(),
                                  true );
        }
    }
    if( widget == NULL ) {
        model::gui::Canvas* canvas = dynamic_cast<model::gui::Canvas*>( root );
        if( canvas != NULL ) {
            widget = addCanvas( canvas, parent );
        }
    }
    if( widget == NULL ) {
        model::gui::Label* label = dynamic_cast<model::gui::Label*>( root );
        if( label != NULL ) {
            widget = new QLabel( parent );
            new StringController( widget,
                                  m_model,
                                  label->key(),
                                  label->showValue() );
        }
    }
    if( widget == NULL ) {
        model::gui::ComboBox* combobox = dynamic_cast<model::gui::ComboBox*>( root );
        if( combobox != NULL ) {
            widget = addCombobox( combobox, parent);
        }
    }
    if( widget == NULL ) {
        model::gui::ElementGroup* element_group = dynamic_cast<model::gui::ElementGroup*>( root );
        if( element_group != NULL ) {
            widget = addElementGroup( element_group, parent);
        }
    }
    if( widget == NULL ) {
        model::gui::RadioButtons* radio_buttons = dynamic_cast<model::gui::RadioButtons*>( root );
        if( radio_buttons != NULL ) {
            widget = addRadiobuttons( radio_buttons, parent);
        }
    }
    if( widget == NULL ) {
        model::gui::SpinBox* spin_box = dynamic_cast<model::gui::SpinBox*>( root );
        if( spin_box != NULL ) {
            widget = addSpinBox( spin_box, parent);
        }
    }
    if( widget == NULL ) {
        model::gui::CheckBox* check_box = dynamic_cast<model::gui::CheckBox*>( root );
        if( check_box != NULL ) {
            widget = addCheckBox( check_box, parent);
        }
    }
    if( widget == NULL ) {
        model::gui::Button* button = dynamic_cast<model::gui::Button*>( root );
        if( button != NULL ) {
            widget = addButton( button, parent);
        }
    }
    if( widget == NULL ) {
        model::gui::HorizontalSlider* horizontal_slider = dynamic_cast<model::gui::HorizontalSlider*>( root );
        if( horizontal_slider != NULL ) {
            widget = addHorizontalSlider( horizontal_slider, parent);
        }
    }
    if( widget == NULL ) {
        model::gui::HorizontalLayout* horizontal_layout = dynamic_cast<model::gui::HorizontalLayout*>( root );
        if( horizontal_layout != NULL ) {
            widget = addHorizontalLayout( horizontal_layout, parent);
        }
    }
    if( widget == NULL ) {
        model::gui::VerticalLayout* vertical_layout = dynamic_cast<model::gui::VerticalLayout*>( root );
        if( vertical_layout != NULL ) {
            widget = addVerticalLayout( vertical_layout, parent);
        }
    }
    if( widget == NULL ) {
        model::gui::DoubleSpinBox* double_spinbox = dynamic_cast<model::gui::DoubleSpinBox*>( root );
        if( double_spinbox != NULL ) {
            widget = addDoubleSpinBox( double_spinbox, parent);
        }
    }
    if( widget == NULL ) {
        model::gui::PopupButton* popup_button = dynamic_cast<model::gui::PopupButton*>( root );
        if( popup_button != NULL ) {
            widget = addPopupButton( popup_button, parent );
        }
    }
    if(widget == NULL ) {
       model::gui::FileDialogButton* file_dialog_button = dynamic_cast<model::gui::FileDialogButton*>( root );
       if(file_dialog_button != NULL)
       {
          widget = addFileDialogButton(file_dialog_button, parent);
       }
    }

    if( widget != NULL ) {

        if( !root->visibilityKey().empty() ) {
            new VisibilityController( widget,
                                      m_model,
                                      root->visibilityKey(),
                                      root->visibilityInverted() );
        }
        if( !root->enabledKey().empty() ) {
            new EnabledController( widget,
                                   m_model,
                                   root->enabledKey(),
                                   root->enabledInverted() );
        }
    }

    return widget;
}

QWidget*
GUIBuilder::addPopupButton( model::gui::PopupButton* root, QWidget* parent )
{
    QString suffix( " " );
    suffix.append( QChar( 0x02ec ) );
    QPushButton* widget = new QPushButton( parent );
    widget->setCheckable( true );
    new impl::StringController( widget, m_model, root->key(),root->showValue(), suffix );

    QWidget* foo = buildGUI( root->child(), widget );
    if( foo != NULL ) {
        foo->setObjectName( "popup" );
        foo->setStyleSheet( "QWidget#popup { border: 1px solid black }");
        new tinia::qtobserver::impl::PopupEventFilter( foo, widget );
    }
    return widget;
}



QWidget*
GUIBuilder::addTabLayout(model::gui::TabLayout *root, QWidget *parent)
{
    QTabWidget* tabWidget = new QTabWidget(parent);

    for( size_t i=0; i<root->children(); i++ ) {
        model::gui::Tab* tab = root->child( i );
        if( tab != NULL ) {
            QWidget* child = buildGUI( tab->child(), tabWidget );
            if( child != NULL ) {
                tabWidget->addTab( child,
                                   QString(prettyName(tab->key(), m_model).c_str()) );
            }
            else {
                std::cerr << __FILE__ << '@' << __LINE__<< ": ERROR: got nullptr.\n";
            }
        }
        else {
            std::cerr << __FILE__ << '@' << __LINE__<< ": ERROR: child is not tab.\n";
        }
    }
   return tabWidget;
}


QWidget*
GUIBuilder::addHorizontalLayout(model::gui::HorizontalLayout *root, QWidget *parent)
{
    QWidget* widget = new QWidget(parent);
    doHorizontalLayout( root, widget );
    return widget;
}

QWidget*
GUIBuilder::addVerticalLayout(model::gui::VerticalLayout *root, QWidget *parent)
{
    QWidget* widget = new QWidget(parent);
    doVerticalLayout( root, widget );
    return widget;
}

void
GUIBuilder::doHorizontalLayout(model::gui::HorizontalLayout *root, QWidget *widget)
{
    QHBoxLayout* layout = new QHBoxLayout( widget );
    widget->setLayout( layout );
    widget->setSizePolicy( QSizePolicy::Minimum, QSizePolicy::Minimum );
    addChildren( static_cast<model::gui::Container1D<model::gui::Element>*>( root ),
                 widget,
                 layout );
}

void
GUIBuilder::doVerticalLayout(model::gui::VerticalLayout *root, QWidget *widget)
{
    QVBoxLayout* layout = new QVBoxLayout( widget );
    widget->setLayout( layout );
    widget->setSizePolicy( QSizePolicy::Minimum, QSizePolicy::Minimum );
    addChildren( static_cast<model::gui::Container1D<model::gui::Element>*>( root ),
                 widget,
                 layout );
}






QLayoutItem*
GUIBuilder::buildLayoutItem( model::gui::Element* root )
{
    model::gui::VerticalSpace* vspace = dynamic_cast<model::gui::VerticalSpace*>( root );
    if( vspace != NULL ) {
        return new QSpacerItem(0, 10, QSizePolicy::Minimum, QSizePolicy::Minimum);
    }
    model::gui::VerticalExpandingSpace* vexspace = dynamic_cast<model::gui::VerticalExpandingSpace*>( root );
    if( vexspace != NULL ) {
        return new QSpacerItem(0, 0, QSizePolicy::Minimum, QSizePolicy::MinimumExpanding);
    }

    model::gui::HorizontalSpace* hspace = dynamic_cast<model::gui::HorizontalSpace*>( root );
    if( hspace != NULL ) {
        return new QSpacerItem(10, 0, QSizePolicy::Minimum, QSizePolicy::Minimum);
    }
    model::gui::HorizontalExpandingSpace* hexspace = dynamic_cast<model::gui::HorizontalExpandingSpace*>( root );
    if( hexspace != NULL ) {
        return new QSpacerItem(0, 0, QSizePolicy::MinimumExpanding, QSizePolicy::Minimum );
    }
    return NULL;
}


QWidget*
GUIBuilder::addGrid(model::gui::Grid *root, QWidget *parent)
{
    QWidget* grid_widget = new QWidget(parent);
    doGridLayout( root, grid_widget );
    return grid_widget;
}

void
GUIBuilder::doGridLayout( model::gui::Grid* root, QWidget* widget )
{
    QGridLayout* grid_layout = new QGridLayout;
    //grid_layout->setHorizontalSpacing(0);
    //grid_layout->setVerticalSpacing(0);
    //grid_layout->setContentsMargins( 0, 0, 0, 0 );
    widget->setLayout( grid_layout );
    addChildren( static_cast<model::gui::Container2D<model::gui::Element>*>( root ),
                 widget,
                 grid_layout );

}



QWidget*
GUIBuilder::addElementGroup(model::gui::ElementGroup *root, QWidget *parent)
{
    QGroupBox* widget = new QGroupBox( parent );
    widget->setFlat( true );

    const std::string key = root->key();
    if( root->showLabel() && !root->key().empty() ) {
        widget->setTitle( (prettyName( root->key(), m_model ) ).c_str() );
    }

    model::gui::Element* child = root->child();
    if( child != NULL ) {
        if( child->type() == model::gui::GRID ) {
            doGridLayout( static_cast<model::gui::Grid*>( child ), widget );
        }
        else if( child->type() == model::gui::HORIZONTAL_LAYOUT ) {
            doHorizontalLayout( static_cast<model::gui::HorizontalLayout*>( child ), widget );
        }
        else if( child->type() == model::gui::VERTICAL_LAYOUT ) {
            doVerticalLayout( static_cast<model::gui::VerticalLayout*>( child ), widget );
        }
        else {
            QHBoxLayout* layout = new QHBoxLayout( widget );
            widget->setLayout( layout );
            addChildren( static_cast<model::gui::Container0D<model::gui::Element>*>( root ),
                         widget,
                         layout );
        }
    }
    return widget;
}


QWidget*
GUIBuilder::addCanvas(model::gui::Canvas *root, QWidget *parent )
{
    QWidget* wrapper = NULL;

    if( m_root_context == NULL ) {
        throw std::runtime_error( "Only GL jobs can create canvases." );
    }

    if( m_observer->perfMode() || m_observer->renderListMode() ) {
        wrapper = new QWidget( parent );

        QWidget* row = new QWidget( wrapper );
        QHBoxLayout* row_layout = new QHBoxLayout( row );
        row->setLayout( row_layout );
        row->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Minimum);
        row_layout->setContentsMargins( 0, 0, 0, 0 );

        tinia::qtobserver::impl::Canvas * canvas = new tinia::qtobserver::impl::Canvas( static_cast<jobobserver::OpenGLJob*>(m_job),
                                      root->key(),
                                      root->boundingBoxKey(),
                                      root->resetViewKey(),
                                      m_model,
                                      wrapper,
                                      m_root_context,
                                      true );
        canvas->setMinimumSize(640, 360);
        canvas->setMaximumSize(2000, 2000);
        canvas->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);

        if( m_observer->renderListMode() ) {
            QComboBox* mode = new QComboBox( row );
            mode->addItem( "Native Rendering" );
            mode->addItem( "Render List" );
            row->layout()->addWidget( mode );
            QObject::connect( mode, SIGNAL(currentIndexChanged(int)),
                              canvas, SLOT(setRenderMode(int)) );
        }
        if( m_observer->perfMode() ) {
            QLabel* fps_label = new QLabel( row );
            fps_label->setText( "fps" );
            QObject::connect( canvas, SIGNAL(updateFPS(QString)), fps_label, SLOT(setText(QString)));
            row->layout()->addWidget(fps_label);
        }
        row_layout->addStretch( 1 );

        QVBoxLayout* wrapper_layout = new QVBoxLayout(wrapper);
        wrapper->setLayout( wrapper_layout );
        wrapper_layout->setSpacing( 0 );
        wrapper->setContentsMargins( 0, 0, 0, 0 );

        wrapper->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
        wrapper->layout()->addWidget(canvas );
        wrapper->layout()->addWidget( row );
    }
    else {
        tinia::qtobserver::impl::Canvas * canvas = new tinia::qtobserver::impl::Canvas( static_cast<jobobserver::OpenGLJob*>(m_job),
                                      root->key(),
                                      root->boundingBoxKey(),
                                      root->resetViewKey(),
                                      m_model,
                                      parent,
                                      m_root_context,
                                      false );
        canvas->setMinimumSize(640, 360);
        canvas->setMaximumSize(2000, 2000);
        canvas->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
        wrapper = canvas;
    }
    return wrapper;
}


QWidget*
GUIBuilder::addCombobox(model::gui::ComboBox *root, QWidget *parent)
{
    QWidget* w = new ComboBox(root->key(), m_model, parent);
    w->setSizePolicy( QSizePolicy::Minimum, QSizePolicy::Minimum );
    return w;
}

QWidget*
GUIBuilder::addRadiobuttons(model::gui::RadioButtons *root, QWidget *parent)
{
    QWidget* w = new RadioButtonGroup(root->key(), m_model,  root->horizontal(), parent);
    w->setSizePolicy( QSizePolicy::Minimum, QSizePolicy::Minimum );
    return w;
}


QWidget*
GUIBuilder::addSpinBox(model::gui::SpinBox *root, QWidget *parent)
{
   QWidget* w = new SpinBox(root->key(), m_model, parent);
   w->setSizePolicy( QSizePolicy::Fixed, QSizePolicy::Minimum );
   return w;
}

QWidget*
GUIBuilder::addCheckBox(model::gui::CheckBox *root, QWidget *parent)
{
   return new CheckBox(root->key(), m_model, parent);
}

QWidget*
GUIBuilder::addButton(model::gui::Button *root, QWidget *parent)
{
    return new Button(root->key(), m_model, parent);
}

QWidget*
GUIBuilder::addHorizontalSlider(model::gui::HorizontalSlider *root, QWidget *parent)
{
   return new HorizontalSlider(root->key(), root->withButtons(),
                                          m_model, parent);
}



QWidget*
GUIBuilder::addDoubleSpinBox(model::gui::DoubleSpinBox *root, QWidget *parent)
{
    return new DoubleSpinBox(root->key(), m_model, parent);
}

void
GUIBuilder::addChildren( model::gui::Container0D<model::gui::Element>*  container,
                         QWidget*                                               widget,
                         QBoxLayout*                                            layout )
{
    model::gui::Element* child = container->child();
    QWidget* child_widget = buildGUI( child, widget );
    if( child_widget != NULL ) {
        if( child->type() == model::gui::CANVAS ) {
            layout->addWidget( child_widget, 1 );
        }
        else {
            layout->addWidget( child_widget );
        }
    }
    else {
        std::cerr << __FILE__ << '@' << __LINE__<< ": ERROR: got nullptr.\n";
    }
}

void
GUIBuilder::addChildren( model::gui::Container1D<model::gui::Element>*  container,
                         QWidget*                                               widget,
                         QBoxLayout*                                            layout )
{
    for( size_t i=0; i<container->children(); i++ ) {
        model::gui::Element* child = container->child( i );
        if( child == NULL ) {
            continue;
        }
        QLayoutItem* layout_item = buildLayoutItem( child );
        if( layout_item != NULL ) {
            layout->addItem( layout_item );
            continue;
        }
        QWidget* child_widget = buildGUI( child, widget );
        if( child_widget != NULL ) {
            if( child->type() == model::gui::CANVAS ) {
                layout->addWidget( child_widget, 1 );
            }
            else {
                layout->addWidget( child_widget );
            }
        }
        else {
            std::cerr << __FILE__ << '@' << __LINE__<< ": ERROR: got nullptr.\n";
        }
    }
}

void
GUIBuilder::addChildren( model::gui::Container2D<model::gui::Element>* container,
                         QWidget* widget,
                         QGridLayout* layout )
{
    for(size_t row = 0; row < container->height(); row++) {
       for(size_t col = 0; col  < container->width(); col++) {

           model::gui::Element* child = container->child(row, col);
           if( child == NULL ) {
               continue;
           }
           QLayoutItem* layout_item = buildLayoutItem( child );
           if( layout_item != NULL ) {
               layout->addItem( layout_item, row, col );
               continue;
           }

           QWidget* child_widget = buildGUI( child, widget );
           if( child_widget != NULL ) {
               if( child->type() == model::gui::CANVAS ) {
                   layout->setColumnStretch( col, 1 );
                   layout->setRowStretch( row, 1 );
               }
               else if( child->type() == model::gui::GRID ) {
                   child_widget->layout()->setContentsMargins(0, 0, 0, 0);
               }
               else if( child->type() == model::gui::VERTICAL_LAYOUT ) {
                   child_widget->layout()->setContentsMargins(0, 0, 0, 0);
               }
               else if( child->type() == model::gui::HORIZONTAL_LAYOUT ) {
                   child_widget->layout()->setContentsMargins(0, 0, 0, 0);
               }
               layout->addWidget( child_widget, row, col );
           }
           else {
               std::cerr << __FILE__ << '@' << __LINE__<< ": ERROR: got nullptr.\n";
           }
       }
    }
}

QWidget * GUIBuilder::addFileDialogButton(model::gui::FileDialogButton *root, QWidget *parent)
{

   FileDialogButton *widget =  new FileDialogButton(root->key(), root->showValue(),
                                                    m_model, parent);
   return widget;
}


}
} // of namespace tinia
